src/transformer.ts

```diff
@@ -1,193 +1,581 @@
+ * expressions using namespaces not explicitly defined on the document (most
+ * commonly `xmlns`).
+ *
+ * TODO (2023-01-18): this logic may be useful downstream, particularly to
+ * address issues with the behavior of {@link correctModelNamespaces}. I'm
+ * exporting it now, as an opportunity to explore that after this is merged.
+ * Maybe we should also document it?
+ */
+export const getNamespaceResolver = (doc: Document) => {
+    const cached = namespaceResolverCache.get(doc);
+
+    if (cached != null) {
+        return cached;
+    }
+
+    const baseResolver = doc.createNSResolver(doc) as NamespaceResolver;
+    const defaultPrefix = doc.lookupPrefix(NAMESPACES.xmlns);
+    const resolver = {
+        lookupNamespaceURI(prefix: string) {
+            if (
+                prefix === defaultPrefix ||
+                (defaultPrefix == null && prefix === 'xmlns')
+            ) {
+                return NAMESPACES.xmlns;
+            }
+
+            return (
+                baseResolver.lookupNamespaceURI(prefix) ??
+                NAMESPACES[prefix as keyof typeof NAMESPACES]
+            );
+        },
+    };
+
+    namespaceResolverCache.set(doc, resolver);
+
+    return resolver;
+};
+
+const getNodesByXPathExpression = <T extends Node = Element>(
+    doc: XMLDocument,
+    expression: string,
+    context: Element = doc.documentElement
+): T[] => {
+    const results: T[] = [];
+    const namespaceResolver = getNamespaceResolver(doc);
+    const result = doc.evaluate(
+        expression,
+        context,
+        namespaceResolver,
+        XPathResult.ORDERED_NODE_SNAPSHOT_TYPE
+    );
+
+    for (let i = 0; i < result.snapshotLength; i += 1) {
+        results.push(result.snapshotItem(i) as T);
+    }
+
+    return results;
+};
+
 const processBinaryDefaults = (
-    doc: XMLJSDocument,
+    doc: Document,
     mediaMap: Record<string, string>
 ) => {
-    doc.find(
-        '/h:html/h:head/xmlns:model/xmlns:bind[@type="binary"]',
-        NAMESPACES
-    ).forEach((bind) => {
-        const nodeset = bind.attr('nodeset');
-
-        if (nodeset && nodeset.value()) {
-            const path = `/h:html/h:head/xmlns:model/xmlns:instance${nodeset
-                .value()
-                ?.replace(/\//g, '/xmlns:')}`;
-            const dataNode = doc.get(path, NAMESPACES);
-
-            if (dataNode) {
-                const text = dataNode.text();
-
-                // Very crude URL checker which is fine for now,
-                // because at this point we don't expect anything other than jr://
-                if (/^[a-zA-Z]+:\/\//.test(text)) {
-                    const value = getMediaPath(mediaMap, text);
-                    const escapedText = escapeURLPath(text);
-
-                    dataNode.attr({ src: value }).text(escapedText);
+    const binaryBindings = Array.from(
+        doc.querySelectorAll(':root > head > model > bind[type="binary"]')
```

Comment:
I've discussed this with @lognaturel, and it may be more obvious in #159 (closed in favor of this PR completing the transition): there is some risk here of losing namespacing. The reason this and several other extant functions switched from XPath to CSS queries is that part of the incremental work (while still running on Node) switched from using `libxmljs` APIs to `linkedom`, because there wasn't an equivalent and sufficiently featureful XPath evaluator with DOM APIs available. (In hindsight, I wish I had considering writing the above `getNodesByXPathExpression` wrapper function sooner, as it would have obviated the need for any such intermediate step.)

All of that is to say, I'd definitely be open to changing these CSS queries back to XPath, I just deferred that in favor of raising the point/question in review.

Reply:
In the spirit of keeping things as similar as possible, my preference would be to stick to XPath queries for now.